### PR TITLE
Reworked error output to send IETF RFC-7807 compatible problem responses

### LINKED ISSUE #14: Use wharf-core pkg/ginutil for error responses
Change error responses to make use of iver-wharf/wharf-core#9

### PR DESCRIPTION
**Pull request by Alexamakans**
Closes #14
Depends on https://github.com/iver-wharf/wharf-core/pull/17

### CODE REVIEW

**Review by fredx30 — APPROVED** [review]

**Review by applejag — CHANGES_REQUESTED** [review]
Wrote a lot of suggestions over at https://github.com/iver-wharf/wharf-provider-azuredevops/pull/14

Instead of repeating myself I suggest looking at those comments to use as template for this one

File: CHANGELOG.md [hunk 1]
```diff
@@ -22,6 +27,9 @@ This project tries to follow [SemVer 2.0.0](https://semver.org/).
 - Added Swagger spec metadata such as version that equals the version of the
   API, contact information, and license. (#4)
 
+- Changed version of Alpine Docker image used as the base image from 3.13.4
+  -> 3.13.5. (#13)
+
```
Comment (applejag) [inline CHANGELOG.md]: Looks like a merge conflict resolution error.

This note was removed in favor of the one below.

```suggestion
```

File: import.go [hunk 2]
```diff
@@ -17,55 +19,78 @@ import (
...
 // @Router /gitlab [post]
 func runGitLabHandler(c *gin.Context) {
 	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
 
 	i := Import{}
 	err := c.BindJSON(&i)
-	if handleError(c, err) {
+	if err != nil {
+		ginutil.WriteInvalidBindError(c, err,
+			"One or more parameters failed to parse when reading the request body for GitHub projects import/refresh")
```
Comment (applejag) [inline import.go]: Need to use `ShouldBindJSON` when using the `ginutil.Write...` functions

https://iver-wharf.github.io/#/development/gin-shouldbind-vs-bind

File: import.go [hunk 3]
```diff
@@ -17,55 +19,78 @@ import (
...
+	if err != nil {
+		if !handleIfAuthError(c, err) {
+			ginutil.WriteProblemError(c, err, problem.Response{
+				Type: "prob/provider/gitlab/creating-importer-error",
+				Title: "Error creating GitLab importer.",
+				Status: http.StatusBadRequest,
+				Detail: "Creation of GitLab importer failed. Check your settings and make sure they are correct.",
+			})
+		}
+
```
Comment (applejag) [inline import.go]: Looking at the implementation, this ought to be an API read error, right?

Suggest moving this problem handling into the newGitLabImporter, similar to how you've done it in the azuredevops PR https://github.com/iver-wharf/wharf-provider-azuredevops/pull/14 with the `WritesError` functions there.

**Review by applejag — CHANGES_REQUESTED** [review]

File: import.go (hunk 3 above)
Comment (Alexamakans) [reply import.go]: It's both Read and Write.
Will move the writing of the problems to the functions called in newGitLabImporter 👍🏻

File: import.go [hunk 4]
```diff
@@ -76,108 +66,113 @@ func runGitLabHandler(c *gin.Context) {
...
-			return wharfapi.Token{}, err
+		if err != nil {
+			ginutil.WriteAPIClientReadError(c, err,
+				fmt.Sprintf(
+					"Unable to get token by id %v. Likely because of a failed request or malformed response.",
+					importData.TokenID))
+		} else if token.TokenID == 0 {
+			err = fmt.Errorf("token with id %v not found", importData.TokenID)
+			ginutil.WriteAPIClientReadError(c, err,
+				fmt.Sprintf("Token with id %v not found", importData.TokenID))
```
Comment (applejag) [inline import.go]: nit:

```suggestion
				fmt.Sprintf("Token with ID %d not found.", importData.TokenID))
```

File: import.go [hunk 5]
```diff
@@ -76,108 +66,113 @@ func runGitLabHandler(c *gin.Context) {
...
+	return token, true
 }
 
-func obtainProvider(wharfClient *wharfapi.Client, tokenID uint, importData *Import) (wharfapi.Provider, error) {
+func obtainProviderWritesProblem(c *gin.Context, wharfClient *wharfapi.Client, tokenID uint, importData *Import) (wharfapi.Provider, bool) {
 	if importData.ProviderID != 0 {
 		provider, err := wharfClient.GetProviderById(importData.ProviderID)
 		if err != nil || provider.ProviderID == 0 {
+			ginutil.WriteAPIClientReadError(c, err,
+				fmt.Sprintf("Unable to get provider by id %d.", importData.ProjectID))
```
Comment (applejag) [inline import.go]: nit:

```suggestion
				fmt.Sprintf("Unable to get provider by ID %d.", importData.ProjectID))
```

File: import.go [hunk 6]
```diff
@@ -76,108 +66,113 @@ func runGitLabHandler(c *gin.Context) {
...
-		if err != nil || token.TokenID == 0 {
-			log.WithError(err).Errorln("unable to get token")
-			return wharfapi.Token{}, err
+		if err != nil {
+			ginutil.WriteAPIClientReadError(c, err,
+				fmt.Sprintf(
+					"Unable to get token by id %v. Likely because of a failed request or malformed response.",
+					importData.TokenID))
+		} else if token.TokenID == 0 {
+			err = fmt.Errorf("token with id %v not found", importData.TokenID)
```
Comment (applejag) [inline import.go]: nit:

```suggestion
			err = fmt.Errorf("token with ID %d not found", importData.TokenID)
```

File: import.go [hunk 7]
```diff
@@ -76,108 +66,113 @@ func runGitLabHandler(c *gin.Context) {
...
+func obtainTokenWritesProblem(c *gin.Context, wharfClient *wharfapi.Client, importData *Import) (wharfapi.Token, bool) {
 	if importData.TokenID != 0 {
 		token, err := wharfClient.GetTokenById(importData.TokenID)
-		if err != nil || token.TokenID == 0 {
-			log.WithError(err).Errorln("unable to get token")
-			return wharfapi.Token{}, err
+		if err != nil {
+			ginutil.WriteAPIClientReadError(c, err,
+				fmt.Sprintf(
+					"Unable to get token by id %v. Likely because of a failed request or malformed response.",
```
Comment (applejag) [inline import.go]: nit:

```suggestion
					"Unable to get token by ID %d. Likely because of a failed request or malformed response.",
```

**Review by applejag — APPROVED** [review]
Just some nits. Otherwise: LGTM!

**Pull request merged**